### ISSUE TITLE
Feature Request: Buck Handle Different Platforms on Python 

### ISSUE THREAD

**Issue by paulcavallaro** [facebook/buck#1206]
# Buck Handle Different Platforms (Mac OS X, Linux, etc.) for Python

Currently I have a combination of `remote_file` + `prebuilt_python_library` rules to pull in various third party python libraries. However sometimes various python libraries contain C extension code and it becomes necessary to build them for a specific platform, in my case Mac OS X and Linux.

I currently don't see a way to have one logical python library rule work both in Mac OS X and Linux.

Concretely:

```BUCK<third_party>
prebuilt_python_library(
    name='faulthandler-macosx_10_10_x86_64',
    binary_src='faulthandler-2.4-cp27-none-macosx_10_10_x86_64.whl',
)

python_library(
    name='faulthandler',
    srcs=[],
    deps=[
        # Would be nice to have some way to include macosx when platform is macosx, and linux if platform is linux here
        ':faulthandler-macosx_10_10_x86_64',
    ],
    visibility=[
        'PUBLIC',
    ],
)
```

You could also imagine having a remote_file properly pull down the built wheel from pypi or a hosted devpi instance. But I still need a way to properly distinguish which of the libraries I should use.

I'm not 100% familiar with Buck, and I know sometimes the documentation is out of date with the implementation, so I thought I'd file a bug/feature request here and see if anyone has any ideas.

Thanks!

**Comment by paulcavallaro** [comment]
So doing some more digging I discovered macros and defining your own rules. Now I have something like:

```
def multiplatform_python_library(name,
                                 deps=[],
                                 os_platform_deps=[],
                                 **kwargs):
    import distutils.util
    import re
    platform = distutils.util.get_platform().replace('-', '_').replace('.', '_')
    for (regex, os_deps) in os_platform_deps:
        if re.match(regex, platform):
            return python_library(
                name=name,
                deps=deps + os_deps,
                **kwargs)

# usage:
prebuilt_python_library(
    name='faulthandler-cp27-none-macosx_10_10_x86_64',
    binary_src='faulthandler-2.4-cp27-none-macosx_10_10_x86_64.whl',
    deps=[],
)

multiplatform_python_library(
    name='faulthandler',
    os_platform_deps=[
        ('macosx_10_10', [':faulthandler-cp27-none-macosx_10_10_x86_64']),
    ],
    visibility=[
        'PUBLIC',
    ],
)
```

Wondering if this seems reasonable, and if other people have tried this and run into problems. My main concern is that this way isn't really supported, and not sure if it's the correct approach.

It might be nice to build into buck something similar to pant's `python_requirement_library` (http://www.pantsbuild.org/build_dictionary.html) for installing libraries from pypi. The downside is that the builds are no longer quite so hermetic, and BUCK doesn't know the full build graph, and would have to understand python packaging as well.

Any thoughts?

**Comment by evanj** [comment]
I had some success modifying your approach to use the same resolution logic that pip does. This modified rule only requires you list the `.whl` files, and it should automatically pick up the same one that pip would. Unfortunately I'm blocked by #1330, so the resulting pex doesn't actually work.

However, if I can figure out that one, this suggests it would likely be possible to create a `BUCK` generator that could take a list of pip requirements, and output the correct rules and download the corresponding wheels so it could then build correctly on multiple platforms.

```
with allow_unsafe_import():
    import pip.wheel
    import pip.pep425tags

    # Inspired by https://github.com/facebook/buck/issues/1206
    # Pip wheel resolution from pip.wheel.cached_wheel:
    # https://github.com/pypa/pip/blob/master/pip/wheel.py#L133
    def multiplatform_python_library(name, deps=[], wheels=[], **kwargs):
        supported_wheels = []
        for wheel_filename in wheels:
            w = pip.wheel.Wheel(wheel_filename)
            if not w.supported():
                continue
            supported_wheels.append((w.support_index_min(), wheel_filename))
        if len(supported_wheels) == 0:
            raise ValueError('no wheel for "%s" that matches platform "%s"' % (
                name, pip.pep425tags.get_platform()))

        supported_wheels.sort()
        wheel_file = supported_wheels[0][1]
        prebuilt_python_library(
            name=name,
            binary_src=wheel_file,
            deps=deps,
            **kwargs
        )
```
